pennylane/devices/qubit/preprocess.py

```diff
@@ -0,0 +1,275 @@
+    circuits, one per term, of a circuit that terminates in ``expval(H)``,
+    if the underlying device does not support Hamiltonian expectation values,
+    or if the device requires finite shots.
+
+    .. warning::
+
+        This method will be tracked by autodifferentiation libraries,
+        such as Autograd, JAX, TensorFlow, and Torch. Please make sure
+        to use ``qml.math`` for autodiff-agnostic tensor processing
+        if required.
+
+    Args:
+        circuit (.QuantumTape): the circuit to preprocess
+        dev (.Device): the device to execute circuit on
+        finite_shots (bool): whether the execution has finite shots or not
+
+    Returns:
+        tuple[Sequence[.QuantumTape], callable]: Returns a tuple containing
+        the sequence of circuits to be executed, and a post-processing function
+        to be applied to the list of evaluated circuit results.
+    """
+    supports_hamiltonian = _supports_observable(dev, "Hamiltonian")
+    grouping_known = all(
+        obs.grouping_indices is not None
+        for obs in circuit.observables
+        if isinstance(obs, qml.Hamiltonian)
+    )
+    # device property present in braket plugin
+    use_grouping = getattr(dev, "use_grouping", True)
+
+    hamiltonian_in_obs = any(isinstance(obs, qml.Hamiltonian) for obs in circuit.observables)
+    expval_sum_in_obs = any(
+        isinstance(m.obs, qml.Sum) and isinstance(m, ExpectationMP) for m in circuit.measurements
+    )
+
+    is_shadow = any(isinstance(m, ShadowExpvalMP) for m in circuit.measurements)
+
+    hamiltonian_unusable = not supports_hamiltonian or (finite_shots and not is_shadow)
+
+    if hamiltonian_in_obs and (hamiltonian_unusable or (use_grouping and grouping_known)):
+        # If the observable contains a Hamiltonian and the device does not
+        # support Hamiltonians, or if the simulation uses finite shots, or
+        # if the Hamiltonian explicitly specifies an observable grouping,
+        # split tape into multiple tapes of diagonalizable known observables.
+        try:
+            circuits, hamiltonian_fn = qml.transforms.hamiltonian_expand(circuit, group=False)
+        except ValueError as e:
+            raise ValueError(
+                "Can only return the expectation of a single Hamiltonian observable"
+            ) from e
+    elif expval_sum_in_obs and not is_shadow:
+        circuits, hamiltonian_fn = qml.transforms.sum_expand(circuit)
+
+    elif (
+        len(circuit._obs_sharing_wires) > 0  # pylint: disable=protected-access
+        and not hamiltonian_in_obs
+        and all(
+            not isinstance(m, (SampleMP, ProbabilityMP, CountsMP)) for m in circuit.measurements
+        )
+    ):
+        # Check for case of non-commuting terms and that there are no Hamiltonians
+        # TODO: allow for Hamiltonians in list of observables as well.
+        circuits, hamiltonian_fn = qml.transforms.split_non_commuting(circuit)
```

Comment:
We will natively support the execution of non-commuting measurements, so we will not need to do this.